src/Experiment/tests/InterfaceAssemblyTest.cs

```diff
@@ -0,0 +1,217 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.CodeDom.Compiler;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Reflection;
+using System.Reflection.Emit.Experimental;
+using System.Reflection.Metadata;
+using System.Reflection.PortableExecutable;
+using System.Runtime.InteropServices;
+using Xunit;
+
+namespace Experiment.Tests
```

Comment:
The testing methodology here is difficult to maintain and will likely be fragile as more complexity is introduced. Instead, I would consider the following.

Create a shape in C# for the type you want to build. For example, define the following interface in a new file that contains all the types you plan to add to a new assembly. Once the type is defined here you can use Reflection on the C# definition and use the resulting `MetadataLoadContext` assets to confirm the shape is identical. This will also allow you to use the `nameof` operator in the tests and make sure the strings and type details are checked by the compiler.

```csharp
namespace Tests
{
    public interface Interface1
    {
        string Func(int a);
    }
}
```

Usage example in test:
```csharp
const string _interfaceName = $"{nameof(Tests)}.{nameof(Tests.Interface1)}";
```

Reply:
Concrete example:

```csharp
namespace Test
{
    public interface IInterface
    {
        string Func(int a);
    }
}

static MetadataLoadContext CreateTypeInAssembly(Type type)
{
    var asmBuilder = AssemblyBuilder.DefineDynamicAssembly(default, default);
    var modBuilder = asmBuilder.DefineDynamicModule(default);

    var typeBuilder = modBuilder.DefineType($"{type.Namespace}.{type.Name}");
    foreach (MethodInfo mi in type.GetMethods())
    {
        typeBuilder.DefineMethod(mi.Name, mi.Attributes);
    }

    // fields, events, properties, etc.

    return default;
}

static Assembly ReadTypeFromMetadata(MetadataLoadContext cxt);

public static void Main(string[] args)
{
    // Create an assembly and return the MetadataLoadContext of the Assembly on disk.
    MetadataLoadContext cxt = CreateTypeInAssembly(typeof(Test.IInterface));

    // From the MetadataLoadContext, read in the assembly and module.
    Assembly asmFromMetadata = ReadTypeFromMetadata(cxt);
    Module moduleFromMetadata = asmFromMetadata.Modules.First();

    // Validate metadata matches the type defined in the test
    Type typeFromMetadata = moduleFromMetadata.GetType($"{nameof(Test)}.{nameof(Test.IInterface)}");
}
```

Reply:
Will do a separate PR with this new testing methodology.